{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caeb2113-868e-494d-a0ca-7229b3463248",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#################################################################################################\n",
    "# working directory\n",
    "W_DIR=\"/home/gpfs/o_kumpitch/05_LosingArchaea/00_data/01_Pasolli_Madagascar_Ethiopia/\" \n",
    "\n",
    "# input directory\n",
    "#I_DIR=$W_DIR\"Data/\" \n",
    "I_DIR=\"/home/gpfs/o_kumpitch/05_LosingArchaea/00_data/01_Pasolli_Madagascar_Ethiopia/03_batch3/pe/\"\n",
    "# name of the output directory\n",
    "#NAME=\"Results\" \n",
    "NAME=\"000_KRAKEN2/03_batch3/02_UGHH_pe\"\n",
    "\n",
    "#create output directories\n",
    "mkdir $W_DIR$NAME\n",
    "mkdir $W_DIR$NAME'/main_output'\n",
    "mkdir $W_DIR$NAME'/bash_scripts'\n",
    "mkdir $W_DIR$NAME'/std_out'\n",
    "\n",
    "# sbatch scripts directory, main output and standard output/error\n",
    "B_DIR=$W_DIR$NAME\"/bash_scripts/\" \n",
    "OUT_D=$W_DIR$NAME'/main_output/'\n",
    "STD_OUT=$W_DIR$NAME'/std_out/'\n",
    "\n",
    "# setup number of cores\n",
    "thr=12\n",
    "\n",
    "# setup memory\n",
    "mem=50G\n",
    "\n",
    "# setup time\n",
    "time=24:00:00\n",
    "\n",
    "# exclude nodes\n",
    "nodes=sx138,sx037,sx038\n",
    "\n",
    "# source conda env\n",
    "conda_env=\". /home/conda/miniconda3/bin/activate kraken_2.1.2\"\n",
    "\n",
    "#################################################################################################\n",
    "\n",
    "#iterate over input directory\n",
    "for i in $(ls -h $I_DIR); do\n",
    "R1=${i//_2.fastq.gz/_1.fastq.gz}\n",
    "R2=${i//_1.fastq.gz/_2.fastq.gz}\n",
    "string=${i%%/} export string\n",
    "\t# check if .fastq.gz file\n",
    "    if [[ $string == *\"fastq.gz\"* ]]; then\n",
    "        arrIN=(${string//_*.fastq.gz/ })\n",
    "        echo ${arrIN[0]}\n",
    "        rm $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#!/bin/bash\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH -p debug\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH -N 1\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH -n\" $thr >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH -t\" $time >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH --mem\" $mem >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH --exclude=\"$nodes >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH -o $STD_OUT${arrIN[0]}.out\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"#SBATCH -e $STD_OUT${arrIN[0]}.err\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo $conda_env >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        \n",
    "        # edit command accordingly!!!!! Always add appropriate number of threads using parameter $thr\n",
    "        #echo \"metaphlan $I_DIR${arrIN[0]}.fastq --input_type fastq --output_file $OUT_D${arrIN[0]}_taxonomic_profile.tsv --samout $OUT_D${arrIN[0]}.sam --nproc $thr\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        echo \"kraken2 --db /home/conda/int_microbiome/databases/kraken2/kraken2_db_uhgg_v2.0.1/ $I_DIR$R1 $I_DIR$R2 --paired --confidence 0.1 --gzip-compressed --memory-mapping --threads $thr --output $OUT_D${arrIN[0]}_kraken2.out --use-names --report $OUT_D${arrIN[0]}_report.out\" >> $B_DIR${arrIN[0]}\".bash\"\n",
    "        \n",
    "        # check sample.bash script before uncomenting this and submitting job\n",
    "        sbatch $B_DIR${arrIN[0]}\".bash\"\n",
    "    fi\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
